webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts

```diff
@@ -0,0 +1,203 @@
+}
+
+/**
+ * A DelayedDataLoader is an object that can be used to batch requests for fetching objects from the server. Instead of
+ * requesting data immediately, it will wait for an amount of time and then send a request to the server for all of
+ * the data which would've been requested during that time.
+ *
+ * More specifically, when queueForLoading is first called, a timer will be started. Until that timer expires, any other
+ * calls to queueForLoading will have the provided identifiers added to the ones from the initial call. When the timer
+ * finally expires, the request will be sent to the server to fetch that data. After that, the timer will be reset and
+ * the next call to queueForLoading will start a new one.
+ *
+ * DelayedDataLoader is intended to be used for loading data for components which are unaware of each other and may appear
+ * in different places in the UI from each other which could otherwise send repeated requests for the same or similar
+ * data as one another.
+ */
+export class DelayedDataLoader<Identifier> extends DataLoader<Identifier, Promise<unknown>> {
+    private readonly wait: number = -1;
+
+    private timeoutId: number = -1;
+    private timeoutCallbacks = new Set<{
+        identifiers: Set<Identifier>;
+        resolve(): void;
+    }>();
+
+    constructor(args: {
+        fetchBatch: (identifiers: Identifier[]) => Promise<unknown>;
+        maxBatchSize: number;
+        wait: number;
+    }) {
+        super(args);
+
+        this.wait = args.wait;
+    }
+
+    public queueForLoading(identifiersToLoad: Identifier[]): void {
+        super.queueForLoading(identifiersToLoad);
+
+        this.startTimeoutIfNeeded();
+    }
+
+    public queueAndWait(identifiersToLoad: Identifier[]): Promise<void> {
+        return new Promise((resolve) => {
+            super.queueForLoading(identifiersToLoad);
+
+            // Save the callback that will resolve this promise so that the caller of this method can wait for its
+            // data to be loaded
+            this.timeoutCallbacks.add({
+                identifiers: new Set(identifiersToLoad),
+                resolve,
+            });
+
+            this.startTimeoutIfNeeded();
+        });
+    }
+
+    private startTimeoutIfNeeded(): void {
+        if (this.timeoutId !== -1) {
+            return;
+        }
+
+        this.timeoutId = window.setTimeout(() => {
+            // Ensure that timeoutId and timeoutCallbacks are cleared before doing anything async so that any calls to
+            // queueForLoading which come while we're fetching this batch are added to the next batch instead
+            this.timeoutId = -1;
+
+            const {identifiers, moreToLoad} = this.startBatch();
+
+            // Start another timeout if there's still more data to load
+            if (moreToLoad) {
+                this.startTimeoutIfNeeded();
+            }
+
+            this.fetchBatch(identifiers).then(() => this.resolveCompletedCallbacks(identifiers));
```

Comment:
Waiting for a `DataLoader` to load is really tricky because:
1. When requests are batched, you can have multiple promises waiting on a single promise. It's the opposite of `Promise.all` because we have multiple places waiting for a single piece of code to run instead of one piece of code waiting for multiple promises to resolve. We have to hold onto each of those promise's `resolve` functions to be able to archive that. 
2. When IDs are deduped or split across multiple requests due to paging, we need to wait until all of the IDs for a given promise have been loaded. I'm a bit worried about the performance implications of storing the IDs for each request, and that's another reason that I made the default version of `queueForLoading` non-blocking.

I spent a lot of time looking at this to make sure that the `DataLoader`'s state is updated "atomically" by making sure that the JS thread can't be pre-empted between the timeout starting and a batch of IDs being removed from the `DataLoader`'s internal state, so please let me know if you see anything that you're at all suspicious about